californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -310,11 +317,19 @@ public void completed(final Exchange exchange) {
 				} else {
 					// in case an empty ACK was lost
 					KeyMID idByMID = KeyMID.fromOutboundMessage(originRequest);
-					exchangeStore.remove(idByMID);
+					exchangeStore.remove(idByMID, exchange);
 				}
 				KeyToken idByToken = KeyToken.fromOutboundMessage(originRequest);
-				exchangeStore.remove(idByToken);
-				if(!originRequest.getOptions().hasObserve()) {
+				exchangeStore.remove(idByToken, exchange);
+				Request request = exchange.getRequest();
+				if (request != originRequest && null != request &&  null != request.getToken()
+						&& !Arrays.equals(request.getToken(), originRequest.getToken())) {
+					// remove starting request also
+					originRequest = request;
+					idByToken = KeyToken.fromOutboundMessage(originRequest);
+					exchangeStore.remove(idByToken, exchange);
```

Comment:
Until that remove of the confusing currentRequest/request, it seems to be dangerous, not the cleanup for both request. By the way, thought the Matcher registers for all sent request, I think changing the request in the exchange is dangerous, because the cleanup may be not execute.
Currently this seems only working, because the token is not changed and the MID is cleaned-up by the ACK/Response.
So my view point is still: the Matcher stuff adds, then it should cleanup also. If Exchange is offering two requests, both had to be cleaned-up. Once, when one of the requests are removed, then this second clean up would be easy to remove.